crypto/s2n_certificate.c

```diff
@@ -297,15 +297,21 @@ int s2n_cert_chain_and_key_load_cns(struct s2n_cert_chain_and_key *chain_and_key
         /* We need to try and decode the CN since it may be encoded as unicode with a
          * direct ASCII equivalent. Any non ASCII bytes in the string will fail later when we
          * actually compare hostnames.
+         * 
+         * `ASN1_STRING_to_UTF8` allocates in both the success case and in the zero return case, but
+         * not in the failure case (negative return value). Therefore, we use `ZERO_TO_DISABLE_DEFER_CLEANUP`
+         * in the failure case to prevent double-freeing `utf8_str`. For the zero and success cases, `utf8_str`
+         * will be freed by the `DEFER_CLEANUP`.
          */
         DEFER_CLEANUP(unsigned char *utf8_str, OPENSSL_free_pointer);
         const int utf8_out_len = ASN1_STRING_to_UTF8(&utf8_str, asn1_str);
         if (utf8_out_len < 0) {
             /* On failure, ASN1_STRING_to_UTF8 does not allocate any memory */
+            ZERO_TO_DISABLE_DEFER_CLEANUP(utf8_str);
             continue;
         } else if (utf8_out_len == 0) {
-            /* We still need to free memory here see https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-7521 */
-            OPENSSL_free(utf8_str);
+            /* We still need to free memory for this case, so let the DEFER_CLEANUP free it
+             see https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-7521 */
```

Comment:
Did we find any useful information from this CVE link? I'm still not sure how it's relevant to this behavior.

Reply:
I'm guessing the references on that CVE provided the missing context, but they are now broken links. There is some context here: https://security.archlinux.org/CVE-2017-7521:
```
 The issue is caused by extract_x509_extension() not checking the return value of 
ASN1_STRING_to_UTF8(), and using then freeing a memory allocation that has 
already been freed if it failed. 
```

Reply:
I'm asking because we may need to update / delete this comment if the link isn't helpful. You did write a more detailed comment up above.

Reply:
It was helpful in the sense that googling for `CVE-2017-7521 ASN1_STRING_to_UTF8` found other links (like the one above) mentioning that the memory should not be freed on failure. I could put that other link in the comment as well